### PR TITLE
Limit PublicAPI

### LINKED ISSUE #1299: Make all public types in /src as internal
# Feature Request

**Are you requesting automatic instrumentation for a framework or library? Please describe.**

Make all public as internal to avoid any issues with IoC and be preapared  for publishing nupkg.

**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**
Consider to use public API analysis like in OTEL SDK.

### PR DESCRIPTION
**Pull request by Kielek**
## Why

Partially fixes #1299

## What

Enable public API analyzer to detect any new publicly available contracts.
Make most of the API public internal.
2 groups of Public API leftovers
 - DuckType related things. Technically it should be possible, but we should adapt changes from the upstream first. Blocked by #402 (I removed the milestone to bring topic on the next SIG meeting).
 - Classed used to defining bytecode instrumentation such as: `LoggingBuilderIntegration`. It should be handled by separate issue. I see two possibilities - find a way how to make it internal, or put it to separate library, and describe that the contract is public due to technical limitations.

## Tests

CI.

## Checklist

- ~~[ ] `CHANGELOG.md` is updated.~~ As we do not have nuget package I think that it is internal change.
- ~~[ ] Documentation is updated.~~
- ~~[ ] New features are covered by tests.~~

### CODE REVIEW

File: src/Directory.Build.props [hunk 1]
```diff
@@ -30,6 +30,19 @@
     <Compile Include="$(MSBuildThisFileDirectory)GlobalSuppressions.cs" Link="GlobalSuppressions.src.cs" />
 
     <!-- reference assemblies let us target .NET Framework without the SDK (for example, on non-Windows) -->
-    <PackageReference Include="Microsoft.NETFramework.ReferenceAssemblies" Version="1.0.2" PrivateAssets="all" />
+    <PackageReference Include="Microsoft.NETFramework.ReferenceAssemblies" Version="1.0.3" PrivateAssets="all" />
```
Comment (Kielek) [inline src/Directory.Build.props:33]: Opportunistic bump, as dependabot is handling this file. See #1312

Comment (pellared) [reply src/Directory.Build.props:33]: I want to at least test https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/1314 before merging this PR.

**Comment by pellared** [comment]
> put it to separate library, and describe that the contract is public due to technical limitations.

We could in theory say the same for "this" library 😉 

> find a way how to make it internal

TBH if we do not find it then I am not convinced that this PR is valuable 😢

File: src/Directory.Build.props (hunk 1 above)
Comment (pellared) [reply src/Directory.Build.props:33]: See https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/1315

**Comment by Kielek** [comment]
> TBH if we do not find it then I am not convinced that this PR is valuable 😢

IMO it is. we are making internal most of easy to use thigns -settings, configrations, calling initialization etc.

**Comment by pellared** [comment]
Before merging this PR I would like others to have a look and agree on the next steps. I do not want to delay too much to avoid conflict resolution. I would like to merge it no later than Tuesday morning.

Nice work 👍

File: src/OpenTelemetry.AutoInstrumentation/.publicApi/netcoreapp3.1/PublicAPI.Unshipped.txt [hunk 2]
```diff
@@ -0,0 +1,158 @@
+const OpenTelemetry.AutoInstrumentation.DuckTyping.DuckAttribute.DefaultFlags = System.Reflection.BindingFlags.Instance | System.Reflection.BindingFlags.Static | System.Reflection.BindingFlags.Public | System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.FlattenHierarchy -> System.Reflection.BindingFlags
```
Comment (RassK) [inline src/OpenTelemetry.AutoInstrumentation/.publicApi/netcoreapp3.1/PublicAPI.Unshipped.txt:1]: Quick visual says it might need resorting?

**Review by RassK — APPROVED** [review]
Just one nit above.

File: src/OpenTelemetry.AutoInstrumentation/.publicApi/netcoreapp3.1/PublicAPI.Unshipped.txt (hunk 2 above)
Comment (Kielek) [reply src/OpenTelemetry.AutoInstrumentation/.publicApi/netcoreapp3.1/PublicAPI.Unshipped.txt:1]: Whole file is autogenerated by VisualStudio quick fix :)
The same way was recommended on the .NET OTel contrib side.

File: src/OpenTelemetry.AutoInstrumentation/Properties/AssemblyInfo.NetFramework.cs [hunk 3]
```diff
@@ -18,9 +18,10 @@
 using System.Reflection;
 using System.Runtime.CompilerServices;
 
-[assembly: AssemblyKeyFileAttribute("keypair.snk")]
+[assembly: AssemblyKeyFile("keypair.snk")]
```
Comment (rajkumar-rangaraj) [inline src/OpenTelemetry.AutoInstrumentation/Properties/AssemblyInfo.NetFramework.cs:21]: Out of curiosity, why a change is needed here?

File: test/OpenTelemetry.AutoInstrumentation.Tests/Configuration/SettingsTests.cs [hunk 4]
```diff
@@ -47,7 +47,7 @@ public void Dispose()
     }
 
     [Fact]
-    public void TracerSettings_DefaultValues()
+    internal void TracerSettings_DefaultValues()
```
Comment (rajkumar-rangaraj) [inline test/OpenTelemetry.AutoInstrumentation.Tests/Configuration/SettingsTests.cs:50]: It's not a recommended approach to add the public API based change to test project. OTel SDK does not modify test projects.

**Review by rajkumar-rangaraj — APPROVED** [review]
Nice work!

Approving the PR with few recommendations.
1. Remove changes to test project
2. Update our Release document to call out the entries from `PublicAPI.Unshipped.txt` should be moved to `PublicAPI.Shipped.txt`.

File: src/OpenTelemetry.AutoInstrumentation/Properties/AssemblyInfo.NetFramework.cs (hunk 3 above)
Comment (pellared) [reply src/OpenTelemetry.AutoInstrumentation/Properties/AssemblyInfo.NetFramework.cs:21]: Looks like an opportunistic refactor 😉

Comment (Kielek) [reply src/OpenTelemetry.AutoInstrumentation/Properties/AssemblyInfo.NetFramework.cs:21]: Semi-auto fix by VS/R# recommendation. Not related.

File: test/OpenTelemetry.AutoInstrumentation.Tests/Configuration/SettingsTests.cs (hunk 4 above)
Comment (Kielek) [reply test/OpenTelemetry.AutoInstrumentation.Tests/Configuration/SettingsTests.cs:50]: It was due to Traces/Metrics/LogsExporter. It is internal, so the method cannot be public. I know that I can probably change it to few public methods, but I hope this solution is acceptable.

See:
 `internal void TracesExporter_SupportedValues(string tracesExporter, TracesExporter expectedTracesExporter)`

**Comment by Kielek** [comment]
> 1. Remove changes to test project

It can be hard. I think that current version is good enough.

> 2. Update our Release document to call out the entries from `PublicAPI.Unshipped.txt` should be moved to `PublicAPI.Shipped.txt`.

Done.

File: test/OpenTelemetry.AutoInstrumentation.Tests/Configuration/SettingsTests.cs (hunk 4 above)
Comment (rajkumar-rangaraj) [reply test/OpenTelemetry.AutoInstrumentation.Tests/Configuration/SettingsTests.cs:50]: It's an additional overhead to maintain these for test projects. I don't have strong preference.

**Review by rajkumar-rangaraj — APPROVED** [review]

**Review by pjanotti — APPROVED** [review]

**Review by pellared — APPROVED** [review]

**Pull request merged**